share/getters/shrex.go

```diff
@@ -0,0 +1,87 @@
+package getters
+
+import (
+	"context"
+	"errors"
+	"time"
+
+	"github.com/celestiaorg/celestia-node/share"
+	"github.com/celestiaorg/celestia-node/share/p2p"
+	"github.com/celestiaorg/celestia-node/share/p2p/peers"
+	"github.com/celestiaorg/celestia-node/share/p2p/shrexeds"
+	"github.com/celestiaorg/celestia-node/share/p2p/shrexnd"
+
+	"github.com/celestiaorg/nmt/namespace"
+	"github.com/celestiaorg/rsmt2d"
+)
+
+var _ share.Getter = (*ShrexGetter)(nil)
+
+const MaxRequestDuration = time.Second * 10
+
+// ShrexGetter is a share.Getter that uses the shrex/eds and shrex/nd protocol to retrieve shares.
+type ShrexGetter struct {
+	edsClient *shrexeds.Client
+	ndClient  *shrexnd.Client
+
+	peerManager        *peers.Manager
+	maxRequestDuration time.Duration
+}
+
+func (sg *ShrexGetter) GetShare(ctx context.Context, root *share.Root, row, col int) (share.Share, error) {
+	return nil, errors.New("shrex-getter: GetShare is not supported")
+}
+
+func (sg *ShrexGetter) GetEDS(ctx context.Context, root *share.Root) (*rsmt2d.ExtendedDataSquare, error) {
+	for {
+		peer, setStatus, err := sg.peerManager.Peer(ctx, root.Hash())
+		if err != nil {
+			log.Debugw("couldn't find peer", "datahash", root.String(), "err", err)
```

Comment:
optional, but would be nice to see how long it was waiting for peer before err.

Reply:
will add

Reply:
actually, going to leave it because needing to add time arithmetic is ugly here